rust/src/operations/delete.rs

```diff
@@ -0,0 +1,944 @@
+    /// Time taken to execute the entire operation
+    pub execution_time_ms: u128,
+    /// Time taken to scan the file for matches
+    pub scan_time_ms: u128,
+    /// Time taken to rewrite the matched files
+    pub rewrite_time_ms: u128,
+}
+
+/// Determine which files contain a record that statisfies the predicate
+async fn find_files<'a>(
+    snapshot: &DeltaTableState,
+    store: ObjectStoreRef,
+    schema: &ArrowSchema,
+    file_schema: Arc<ArrowSchema>,
+    candidates: Vec<&'a Add>,
+    state: &SessionState,
+    expression: &Expr,
+) -> DeltaResult<Vec<&'a Add>> {
+    let mut files = Vec::new();
+    let mut candidate_map: HashMap<String, &'a Add> = HashMap::new();
+
+    let table_partition_cols = snapshot
+        .current_metadata()
+        .ok_or(DeltaTableError::NoMetadata)?
+        .partition_columns
+        .clone();
+
+    let mut file_groups: HashMap<Vec<ScalarValue>, Vec<PartitionedFile>> = HashMap::new();
+    for action in candidates {
+        let mut part = partitioned_file_from_action(action, schema);
+        part.partition_values
+            .push(ScalarValue::Utf8(Some(action.path.clone())));
+
+        file_groups
+            .entry(part.partition_values.clone())
+            .or_default()
+            .push(part);
+
+        candidate_map.insert(action.path.to_owned(), action);
+    }
+
+    let mut table_partition_cols = table_partition_cols
+        .iter()
+        .map(|c| {
+            Ok((
+                c.to_owned(),
+                wrap_partition_type_in_dict(schema.field_with_name(c)?.data_type().clone()),
+            ))
+        })
+        .collect::<Result<Vec<_>, ArrowError>>()?;
+    // Append a column called __delta_rs_path to track the file path
+    table_partition_cols.push((PATH_COLUMN.to_owned(), DataType::Utf8));
+
+    let physical_schema = file_schema.clone();
+    let logical_schema: DFSchema = file_schema.as_ref().clone().try_into()?;
+    let execution_props = ExecutionProps::new();
+
+    let predicate_expr = create_physical_expr(
+        expression,
+        &logical_schema,
+        &physical_schema,
+        &execution_props,
+    )?;
+    let parquet_scan = ParquetFormat::new()
+        .create_physical_plan(
+            state,
+            FileScanConfig {
+                object_store_url: store.object_store_url(),
+                file_schema,
+                file_groups: file_groups.into_values().collect(),
+                statistics: snapshot.datafusion_table_statistics(),
+                projection: None,
+                limit: None,
+                table_partition_cols,
+                infinite_source: false,
+                output_ordering: None,
+            },
+            None,
+        )
+        .await?;
+
+    let filter: Arc<dyn ExecutionPlan> =
+        Arc::new(FilterExec::try_new(predicate_expr, parquet_scan.clone())?);
+    let limit: Arc<dyn ExecutionPlan> = Arc::new(LocalLimitExec::new(filter, 1));
```

Comment:
Does this issue a scan to every file? And maybe skips partitions that are irrelevant?

I wonder if we can do this in memory just with the partition columns. This example from DataFusion might be good inspiration:

https://github.com/apache/arrow-datafusion/blob/a38480951f40abce7ee2d5919251a1d1607f1dee/datafusion/core/src/datasource/listing/helpers.rs#L233-L247

Reply:
Pruning is performed based on the predicate but I do agree if only partition columns exist in the predicate then we just do it in memory and would better than relying on pruning being correct.

Reply:
Ah I see the issue. We'll need to make some improvements to the DataFusion expressions so we can handle this better.

Reply:
I implemented the solution to create a memory table and load it with the partition values. I did not map all possible types but if the current approach looks good I'll finish it up.

Reply:
Thanks, this looks pretty good!